File: daily_problems/2024/09/0917/personal_submission/cf433c_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    vector<int> nums(m);
    for (auto &v: nums) cin >> v;

    long long orig = 0;
    vector<long long> cur(n + 1);
    vector<vector<int>> tmp(n + 1);

    for (int i = 1; i < m; i ++) {
        int v = abs(nums[i] - nums[i-1]);
        orig += v;
        
        if (nums[i] != nums[i-1]) {
            cur[nums[i]] += v;
            cur[nums[i-1]] += v;
            tmp[nums[i]].emplace_back(nums[i-1]);
            tmp[nums[i-1]].emplace_back(nums[i]);
        }
    }

    long long res = 0;

    for (int i = 1; i <= n; i ++) {
        if (tmp[i].size()) {
            sort(tmp[i].begin(), tmp[i].end());
            int x = tmp[i][tmp[i].size() / 2];

            long long cur_res = 0;
            for (auto &val: tmp[i]) cur_res += abs(val - x);

            res = max(res, cur[i] - cur_res);
        }
    }
    cout << orig - res;

    return 0;
}

File: daily_problems/2024/09/0917/personal_submission/cf433c_yawn_sean.py
# Submission link: https://codeforces.com/contest/433/submission/281547087
def main():
    n, m = MII()
    nums = LII()

    cur = [0] * (n + 1)
    tmp = [[] for _ in range(n + 1)]

    orig = 0
    for i in range(1, m):
        v = abs(nums[i] - nums[i - 1])
        orig += v

        if nums[i] != nums[i-1]:
            cur[nums[i]] += v
            cur[nums[i-1]] += v
            tmp[nums[i]].append(nums[i-1])
            tmp[nums[i-1]].append(nums[i])

    res = 0
    for i in range(1, n + 1):
        if len(tmp[i]):
            tmp[i].sort()
            x = tmp[i][len(tmp[i]) // 2]
            res = max(res, cur[i] - sum(abs(x - v) for v in tmp[i]))

    print(orig - res)